### PR TITLE
Date range

### PR DESCRIPTION
**Pull request by sabbaka**
Datepicker.

### CODE REVIEW

File: pkg/session_recording/recordings.jsx [hunk 1]
```diff
@@ -238,16 +243,51 @@
...
+    var Datepicker = class extends React.Component {
+        constructor(props) {
+            super(props);
+            this.handleDateChange = this.handleDateChange.bind(this);
+        }
+
+        componentDidMount() {
+            var funcDate = this.handleDateChange;
+            let elemId = '#' + this.props.element;
+            $(elemId).datepicker({
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I think the proper way to do this is React "refs": https://facebook.github.io/react/docs/refs-and-the-dom.html

File: pkg/session_recording/recordings.jsx [hunk 2]
```diff
@@ -425,13 +491,31 @@
...
+                        <div className="content-header-extra">
+                            <table class="form-table-ct">
+                                <tr>
+                                    <td className="top"><label className="control-label" for="dateSince">Date Since</label></td>
+                                    <td><Datepicker element={'dateSince'} date={dateSince} onDateChange={this.handleDateSinceChange} /></td>
+                                    <td className="top"><label className="control-label" for="dateUntil">Date Until</label></td>
+                                    <td><Datepicker element={'dateUntil'} date={dateUntil} onDateChange={this.handleDateUntilChange} /></td>
+                                </tr>
+                            </table>
+                        </div>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I think the markup above should be a part of the RecordingList component, and the hooks should be passed through it to Datepickers. This way the details of the markup could be hidden from View and the RecordingList could better integrate with it.

File: pkg/session_recording/recordings.jsx [hunk 3]
```diff
@@ -364,9 +414,11 @@
          */
         journalctlStart() {
             /* TODO Lookup UID of "tlog" user on module init */
-            var matches = ["_UID=987"];
-            var options = {follow: true, count: "all"};
-
+            /* Or Lookup by tlog-rec parameter */
+            // var matches = ["_COMM=tlog-rec"];
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: This will go away, and we will be using tlog-rec-session only.
First, because Cockpit is a system-level, administration tool, and tlog-rec is an end-user tool, while tlog-rec-session is a system-level tool.
Second, because _COMM can be spoofed by the user, or lost by journald.

File: pkg/session_recording/recordings.jsx [hunk 4]
```diff
@@ -398,6 +450,20 @@
             this.journalctl = null;
         }
 
+        handleDateSinceChange(date) {
+            this.setState({recordingList: []});
+            this.setState({dateSince: date });
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I think it's better to use a single invocation of setState().

File: pkg/session_recording/recordings.jsx [hunk 5]
```diff
@@ -398,6 +450,20 @@
...
+            this.recordingMap = {};
+            this.journalctlStart();
+        }
+
+        handleDateUntilChange(date) {
+            this.setState({recordingList: []});
+            this.setState({dateUntil: date });
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Same here.

File: pkg/session_recording/recordings.jsx [hunk 6]
```diff
@@ -398,6 +450,20 @@
             this.journalctl = null;
         }
 
+        handleDateSinceChange(date) {
+            this.setState({recordingList: []});
+            this.setState({dateSince: date });
+            this.recordingMap = {};
+            this.journalctlStart();
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: You need to stop journalctl first, if it's running. Maybe add journalctlRestart for that.

File: pkg/session_recording/recordings.jsx [hunk 7]
```diff
@@ -398,6 +450,20 @@
...
+        }
+
+        handleDateUntilChange(date) {
+            this.setState({recordingList: []});
+            this.setState({dateUntil: date });
+            this.recordingMap = {};
+            this.journalctlStart();
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Same here.

**Review by spbnick — CHANGES_REQUESTED** [review]
I like the direction you're going, looks good.
I left some comments in the code, though.

File: pkg/session_recording/recordings.jsx [hunk 8]
```diff
@@ -48,7 +51,7 @@
         var d = new Date(ms);
         return (
             padInt(d.getFullYear(), 4) + '-' +
-            padInt(d.getMonth(), 2) + '-' +
+            padInt(d.getMonth() + 1, 2) + '-' +
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Please make this fix into a separate commit.

File: pkg/session_recording/recordings.jsx [hunk 9]
```diff
@@ -222,13 +233,31 @@
...
-                var r = list[i];
-                var columns = [r.user,
+            const dateSince = this.state.dateSince;
+            const dateUntil = this.state.dateUntil;
+            let columnTitles = [_("User"), _("Start"), _("End"), _("Duration")];
+            let list = this.props.list;
+            let rows = [];
+            for (let i = 0; i < list.length; i++) {
+                let r = list[i];
+                let columns = [r.user,
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx:293]: Make a separate commit switching to `let` with a short comment why that's necessary. Also, do this change everywhere in the file, where necessary, in the same commit.

File: pkg/session_recording/recordings.jsx [hunk 10]
```diff
@@ -238,16 +267,68 @@
                             navigateToItem={this.navigateToRecording.bind(this, r)}/>);
             }
             return (
+                <div>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Please indent the contents of this `<div>`.

File: pkg/session_recording/recordings.jsx [hunk 11]
```diff
@@ -238,16 +267,68 @@
...
+                            </td>
+                        </tr>
+                    </table>
+                </div>
                 <Listing.Listing title={_("Sessions")}
                                  columnTitles={columnTitles}
                                  emptyCaption={_("No recorded sessions")}
-                                 fullWidth={false}>
+                                 fullWidth={false}
+                                 >
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Is this kind of closing angle bracket wrapping in line with Cockpit conventions? Should it be on the last attribute's line instead, or should it be on separate line aligned with the opening angle bracket? It also doesn't seem to be consistent with the rest of the code in the file. Please make sure this is consistent.

File: pkg/session_recording/recordings.jsx [hunk 12]
```diff
@@ -269,6 +352,10 @@
                 recordingList: [],
                 /* ID of the recording to display, or null for all */
                 recordingID: cockpit.location.path[0] || null,
+                /* Date since */
+                dateSince: null,
+                /* Date until */
+                dateUntil: null,
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: The two comments above don't add anything to what the property names already say. Please either expand on each property's purpose, or remove the comments.

File: pkg/session_recording/recordings.jsx [hunk 13]
```diff
@@ -398,6 +491,35 @@
...
+            }
+            this.journalctlStart();
+        }
+
+        /*
+         * Handles Datepicker Since.
+         */
+        handleDateSinceChange(date) {
+            this.setState({dateSince: date});
+            this.journalctlRestart();
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: I'm not sure having journalctlRestart here would be correct. First of all, the date doesn't necessarily change, and there might be no update, or it can be queued for later. See https://facebook.github.io/react/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

I would say, we need to try putting the journalctl restart into componentDidUpdate. However, any associated state change needs to be here.

File: pkg/session_recording/recordings.jsx [hunk 14]
```diff
@@ -398,6 +491,35 @@
...
+         * Assumes journalctl is running.
+         */
+        journalctlRestart() {
+            if(this.journalctlIsRunning()) {
+                this.journalctl.stop();
+                this.recordingMap = {};
+                this.setState({recordingList: []});
+            }
+            this.journalctlStart();
+        }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: The function description and code don't match. First of all, the function doesn't assume journalctl is running, it handles both cases: running and not running. Next, it not only restarts journalctl, it also clears state, and that must be mentioned in the description, if you're going to keep that there.

File: pkg/session_recording/recordings.jsx [hunk 15]
```diff
@@ -398,6 +491,35 @@
...
+            this.setState({dateSince: date});
+            this.journalctlRestart();
+        }
+
+        /*
+         * Handles Datepicker Until.
+         */
+        handleDateUntilChange(date) {
+            this.setState({dateUntil: date});
+            this.journalctlRestart();
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Same here.

File: pkg/session_recording/recordings.jsx [hunk 16]
```diff
@@ -425,13 +547,24 @@
...
-                    />
+                    <div>
+                    <RecordingList
+                        dateSince={dateSince} onDateSinceChange={this.handleDateSinceChange}
+                        dateUntil={dateUntil} onDateUntilChange={this.handleDateUntilChange}
+                        list={this.state.recordingList} />
+                    </div>
+                );
+            }
+            else {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Please keep `else` and braces on the same line. It seems to be the prevailing convention in Cockpit.

File: pkg/session_recording/recordings.jsx [hunk 17]
```diff
@@ -425,13 +547,24 @@
...
                 return (
-                    <Recording
-                        recording={this.recordingMap[this.state.recordingID]}
-                    />
+                    <div>
+                    <RecordingList
+                        dateSince={dateSince} onDateSinceChange={this.handleDateSinceChange}
+                        dateUntil={dateUntil} onDateUntilChange={this.handleDateUntilChange}
+                        list={this.state.recordingList} />
+                    </div>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: We don't need this `<div>` anymore, it seems, please remove.

File: pkg/session_recording/recordings.jsx [hunk 18]
```diff
@@ -425,13 +547,24 @@
...
+                );
+            }
+            else {
+                return (
+                    <div>
+                        <Recording recording={this.recordingMap[this.state.recordingID]} />
+                    </div>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Same here.

File: pkg/session_recording/recordings.jsx [hunk 19]
```diff
@@ -238,16 +267,68 @@
...
+                                 fullWidth={false}
+                                 >
                     {rows}
                 </Listing.Listing>
+                </div>
             );
         }
     };
 
+    var Datepicker = class extends React.Component {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Please provide a brief description of this class, similar to others.

File: pkg/session_recording/recordings.jsx [hunk 20]
```diff
@@ -331,6 +418,10 @@
                          j--);
                     recordingList.splice(j + 1, 0, r);
                 } else {
+                    /* Remove existing recording for list if it's not in correct timeframe */
+                    if(ts < r.start) {
+                        delete this.recordingMap[id];
+                    }
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Is this still needed after you added state clearing to date range change?

**Review by spbnick — CHANGES_REQUESTED** [review]
This looks good, but I still have a bunch of comments and requests. Mostly cleanup, but also some logic changes.

Among others, please arrange this into logically-separate commits, with appropriate descriptions. I think we can do this in three commits: fixing the month number issue, switching to `let`, and adding date filtering. Ping me if you need help with git and making those commits.

File: pkg/session_recording/recordings.jsx (hunk 9 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx:293]: Basically, in ES6 it's considered to use let always when you don't need var. This will lock let to scope without possible conflicts with variables on upper levels ( possible problems with closures ). Also, it's more memory efficient - cause let will be destroyed immediately when scope is dead.

Comment (spbnick) [reply pkg/session_recording/recordings.jsx:293]: Thank you. I guessed as much. What I was trying to say was just when you make a separate commit with this change, add a brief justification in the message explaining that. Then that will be a well-formed commit :)

File: pkg/session_recording/recordings.jsx (hunk 20 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: No, it's not needed, I think it was just a mistake in commit. And in addition to this such usage of delete for array element should be avoided, because it will leave like an empty slot, so it was a bad idea to use delete here.

File: pkg/session_recording/recordings.jsx (hunk 14 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: I've moved this to another function.

File: pkg/session_recording/recordings.jsx (hunk 13 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Event on date change won't fire before date is actually changed. I think we can keep it as is for now, but I will think about better way for future.

File: pkg/session_recording/recordings.jsx (hunk 18 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Fixed.

File: pkg/session_recording/recordings.jsx (hunk 16 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Done.

File: pkg/session_recording/recordings.jsx (hunk 17 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Got it.

File: pkg/session_recording/recordings.jsx (hunk 12 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Agree. I've removed them.

File: pkg/session_recording/recordings.jsx (hunk 19 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: I've added description for this also.

File: pkg/session_recording/recordings.jsx (hunk 11 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Just a mistake, sorry.

File: pkg/session_recording/recordings.jsx (hunk 10 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Done.

File: pkg/session_recording/recordings.jsx (hunk 8 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Done as a separate commit. Thank you.

File: pkg/session_recording/recordings.jsx (hunk 9 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx:293]: Done as a separate commit for whole file.

**Comment by sabbaka** [comment]
I've worked through all requested changes and pushed updated, proper commits. Please, review.

File: pkg/session_recording/recordings.jsx [hunk 21]
```diff
@@ -238,17 +259,74 @@
...
             );
         }
     };
 
     /*
+     * A component representing a datepicker based on bootstrap-datepicker.
+     * Requires jQuery and bootstrap-datepicker.
+     * Properties:
+     * - onDateChange: function to call on date change event of datepicker.
+     */
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx:93]: A nice description, thank you.

File: pkg/session_recording/recordings.jsx [hunk 22]
```diff
@@ -238,17 +259,74 @@
...
+    let Datepicker = class extends React.Component {
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx:94]: Even though evidently this works, it still makes me a bit nervous. Could you please move this class above the one that uses it, i.e. RecordingList? Thank you.

File: pkg/session_recording/recordings.jsx [hunk 23]
```diff
@@ -238,17 +259,74 @@
...
-                                 columnTitles={columnTitles}
-                                 emptyCaption={_("No recorded sessions")}
-                                 fullWidth={false}>
-                    {rows}
-                </Listing.Listing>
+                <div>
+                    <div className="content-header-extra">
+                        <table class="form-table-ct">
+                            <tr>
+                                <td className="top"><label className="control-label" for="dateSince">Date Since</label></td>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Please wrap this line on 120 columns boundary to comply with [Cockpit coding guidelines](https://github.com/cockpit-project/cockpit/wiki/Cockpit-Coding-Guidelines).

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Kyrill, functionally and structurally this looks great now, I only have cosmetic comments left. Should be easy to fix.

Apart from inline comments, here are some requests on commit messages:
- Please don't terminate commit subjects with a dot, consider them (article) titles. Those are never terminated with a full stop, because there's nothing following them, same as with commit subjects.
- Please keep commit subjects in the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood), i.e. worded like commands, where possible. I.e. write "Add ..." instead of "Added ..." This not only makes subjects shorter, but also makes it easier to reason about particular commits when merging/reversing/cherry-picking them. The first two commits are perfect examples, but the third one needs to be fixed.
- Keep commit subjects short and to the point, so they're quick to grasp, e.g. when reading a long commit log. Ideally within about 50 characters. It's sometimes a difficult and interesting exercise in English to try to cram the commit meaning in such a short phrase, but it helps a lot. The first two commits are perfectly fine in that regard, but the third one could use improvement. It is not just a bit too long, it's also somewhat difficult to grasp. How about something like "Add date range limiting"?
- By now I know perfectly well what those commits are about, but often, especially when contributing to somebody else's project (or for reading it a year later), it is useful to expand a little on the commit contents and purpose, and justify it in the commit message. Could you please give *short* explanation what was wrong with month display in the first commit message, why you needed to  change `var` to `let` in the second commit message, and, *briefly*, how you implemented date range limiting in the third commit message?

Thank you!

File: pkg/session_recording/recordings.jsx [hunk 24]
```diff
@@ -238,26 +259,85 @@
...
+                                <td>
+                                    <Datepicker date={dateSince} onDateChange={this.handleDateSinceChange} />
+                                </td>
+                                <td className="top"><label className="control-label" for="dateUntil">Date Until</label></td>
```
Comment (spbnick) [inline pkg/session_recording/recordings.jsx]: Oh, and this one too!

File: pkg/session_recording/recordings.jsx (hunk 22 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx:94]: Well yeah, agree for sure.

File: pkg/session_recording/recordings.jsx (hunk 23 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Done.

File: pkg/session_recording/recordings.jsx (hunk 24 above)
Comment (sabbaka) [reply pkg/session_recording/recordings.jsx]: Fixed.

**Comment by sabbaka** [comment]
I've made all requested changes .

**Comment by spbnick** [comment]
Thank you, merged!
